FSF-2669, FSF-2667, FSF-2188 Master state logic

yvesb04 opened the pull request:
various issues fixed, completed state logic actuators

marhcouto commented on master/include/logic/stateLogic.hpp:
```diff
@@ -1,104 +1,109 @@
     }
 }
+
+inline void ASState::performEmergencyOperations() {
```
What is this function for?

yvesb04 replied:
avoiding repeating the emergency operations 3 times, just call the function

marhcouto replied:
but there is only one statement inside of it

marhcouto reviewed (changes requested):
Overall good job! Just missing doccumentation, which is important

yvesb04 commented:
> Overall good job! Just missing doccumentation, which is important

will add

andrelds11 commented on master/platformio.ini:
```diff
@@ -4,15 +4,17 @@ default_envs = teensy41, teensy41-debug
 build_flags = -D EMBEDDED
-board = teensy41
+;board = teensy41
+board = teensy40
```
why the change here, shouldn't it be another environment?

yvesb04 replied:
yes i pushed the wrong thing, for testing we need to use teensy40 as that's the breadboard testing board

andrelds11 commented on master/include/embedded/digitalReceiver.hpp:
```diff
@@ -4,126 +4,133 @@
-      digitalRead(MISSION_TRACKDRIVE_PIN) * TRACKDRIVE |
-      digitalRead(MISSION_EBSTEST_PIN) * EBS_TEST |
-      digitalRead(MISSION_INSPECTION_PIN) * INSPECTION);
+inline void DigitalReceiver::readMission() const {
```
why the consts in these functions? Using it in a read/set function may give the false idea that this is not changing any variable

yvesb04 replied:
clion suggestions, they can be const as they are not editing any class attributes but rather (the pointer adress i.e), but we can remove this

marhcouto commented on master/include/logic/checkupManager.hpp:
```diff
@@ -112,10 +201,20 @@ inline bool CheckupManager::r2dCheckup() const {
 
 inline bool CheckupManager::emergencyCheckup() const {
-    //TODO Continuous monitoring sequence
+    if (_systemData->failureDetection.hasAnyComponentTimedOut()
```
Isn't the EBS check missing? Like the pneumatic sensors?

marhcouto commented on master/include/logic/checkupManager.hpp:
```diff
@@ -112,10 +201,20 @@ inline bool CheckupManager::r2dCheckup() const {
     return EXIT_SUCCESS;
 }
 
+inline bool CheckupManager::drivingCheckup() const {
```
I am confused by this

marhcouto commented on master/include/logic/stateLogic.hpp:
```diff
@@ -1,104 +1,108 @@
-            }
-        // TODO: perform necessary actions to enter AS_OFF
+
+            DigitalSender::enterOffState();
```
The way these if statements with breaks are made are quite confusing

marhcouto commented on master/include/logic/stateLogic.hpp:
```diff
@@ -1,104 +1,108 @@
-            if (_checkupManager.missionFinishedCheckup()) {
-                break;
-            }
+            if (_checkupManager.drivingCheckup()) break;
```
I think the combination of having a successfull exit as a 0 (which is typical in some contexts but may be unexpected in this one, as usually it is coupled with an error check) and the fact that you use the break instead of reversing the logic and puting the rest inside of the if statement leaves this code more confusing than it should be.

marhcouto commented on master/include/embedded/digitalSender.hpp:
```diff
@@ -3,6 +3,132 @@
+    void blinkLED();
 };
+
+inline void DigitalSender::sendDigitalSignal(const int pin, const int signal) {
```
What is this for?

yvesb04 replied:
buzzer signal

marhcouto reviewed (commented):
OVerall great job at making organized code, just a last suggestion: the state machine could be further simplified if you just acted the states functions everytime in their case, instead of doing it once before you enter them

marhcouto reviewed (changes requested):
Well those are the changes requested

marhcouto reviewed (commented):
Sorry for taking so long

yvesb04 commented:
Everything fixed I believe, please re-review. Missing correct brake pressure checks somewhere (probably), please follow the state logic sequence and tell me where so I can fix this. Implemented watchdog correctly too. removed consts, renamed everything for easy to follow logic

andrelds11 commented:
> Everything fixed I believe, please re-review. Missing correct brake pressure checks somewhere (probably), please follow the state logic sequence and tell me where so I can fix this. Implemented watchdog correctly too. removed consts, renamed everything for easy to follow logic

I think the brake pressure is right, your checking it in both manual and as off states

marhcouto commented on master/include/logic/systemDiagnostics.hpp:
```diff
@@ -35,19 +37,19 @@ struct InternalLogics {
+    Metro pcAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
+    Metro steerAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
+    Metro inversorAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
+    Metro bmsAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
```
To remove

marhcouto replied:
Just the bms one

marhcouto commented on master/include/logic/checkupManager.hpp:
```diff
@@ -68,75 +94,179 @@ class CheckupManager {
-    return EXIT_SUCCESS;
+inline bool CheckupManager::shouldEnterEmergency() const {
+    if (_systemData->failureDetection.hasAnyComponentTimedOut() ||
+        _systemData->failureDetection.emergencySignal ||
```
Missing check to pneumatic sensors
```suggestion
        _systemData->digitalData.pneumatic_line_pressure || 
```

marhcouto commented on master/include/logic/checkupManager.hpp:
```diff
@@ -68,75 +94,179 @@ class CheckupManager {
+                state = CheckupState::CLOSE_SDC;
+                //Start toggling watchdog again (will only actually happen when the car is in AS_READY)
+            }
+            break;
```
Needs to start toggling watchdog again for the next states

marhcouto commented on master/include/logic/checkupManager.hpp:
```diff
@@ -68,75 +94,179 @@ class CheckupManager {
+inline bool CheckupManager::shouldEnterEmergency() const {
+    if (_systemData->failureDetection.hasAnyComponentTimedOut() ||
+        _systemData->failureDetection.emergencySignal ||
+        _systemData->sdcState_OPEN ||
```
Needs to check if ASMS is off or if hydraulic brakes are active for AS Driving as well

marhcouto commented on master/include/logic/systemDiagnostics.hpp:
```diff
@@ -35,19 +37,19 @@ struct InternalLogics {
-            bmsAliveTimestamp;
+    Metro pcAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
+    Metro steerAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
+    Metro inversorAliveTimestamp{COMPONENT_TIMESTAMP_TIMEOUT};
```
Missing updates to this timestamp

marhcouto commented on master/include/logic/checkupManager.hpp:
```diff
@@ -68,75 +94,179 @@ class CheckupManager {
-    //TODO Continuous monitoring sequence
-    return EXIT_SUCCESS;
+inline bool CheckupManager::shouldEnterEmergency() const {
+    if (_systemData->failureDetection.hasAnyComponentTimedOut() ||
```
Also needs to check if TS is on (message from bamocar telling the tension)

marhcouto reviewed (changes requested):
Make changes and then merge right away
